### ISSUE TITLE
script data format when redeemers are empty?

### ISSUE THREAD

**Issue by ashisherc** [IntersectMBO/cardano-ledger#2604]
https://github.com/input-output-hk/cardano-ledger/blob/9c3b4737b13b30f71529e76c5330f403165e28a6/eras/alonzo/test-suite/cddl-files/alonzo.cddl#L94

What is the format for constructing script data to get scriptDataHash when there are no redeemers? This is being used by cli which is only adding a datum to the tx, used when embedding a datum to tx with datum hash output, below tx cbor constructed by cli for ex,

```
84a7008182582060ac17c64f40d8182f8528939583515c72a04703abe893af18bf0dee9f715437000d80018282581d603ffa7225c34af125fd4acaf0cf51a9de98b1e1c5df792e599ad525321a0baf0cd583581d704b8980baf5b09f3573f26e100df813dd24a619d78188b9d56471e51f1a001a4ec45820923918e403bf43c34b4ef6b48eb2ee04babed17320d8d1b9ff9ad086e86f44ec021a00029519031a05f5e0ff0e800b58202f50ea2546f8ce020ca45bfcf2abeb02ff18af2283466f888ae489184b3d2d39a200818258201b949a6f5e0eb3c7d4def14a0301d2ef3d543f9f9c318bdacd5d198f3a3b481758408e80f7f92c2d168469f606d06ac27e9e47380b95948476b9a1647f757fe9bff486b676db4905ddfe8cf0773f5b30be89b6e87df5ef1a6f2393553b60b745bc0e0481d87980f5f6
```

I tried combination of excluding the redeemer part, langView part, empty array cbor for redeemer, but nothing seemed to produce a valid scriptDataHash that is produced by the cli.

Thanks,

**Comment by JaredCorduan** [comment]
The answer is sad. :disappointed:. The redeemers are contributing `80` to the serialized data for the hash, which is the CBOR encoding of "list of length 0".

I can guess how this happened. Originally we did not allow extraneous datums of any kind. So if the script integrity hash was present, there must be at least one redeemer. Then we later made the decision to allow extraneous datums in one case, namely the creation of outputs that had a datum hash.

The code here: https://github.com/input-output-hk/cardano-ledger/blob/f91a8760711bb66b5ebfedf7867cec1965c924f3/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs#L244-L246

should instead be:

```
    let dBytes = if nullDats d then mempty else originalBytes d
        mBytes = if nullRedeemers m then mempty else originalBytes m
        lBytes = serializeEncoding' (encodeLangViews l)
     in mBytes <> dBytes <> lBytes
```

**Comment by ashisherc** [comment]
Tried with 80, cbor encoding of []. But that results in a different hash, specifically below is what I am hashing,

redeemerHex = 80
datumHex = 81d87980
lanViewHex = `a141005901d59f1a000302590001011a00060bc719026d00011a000249f01903e800011a000249f018201a0025cea81971f70419744d186419744d186419744d186419744d186419744d186419744d18641864186419744d18641a000249f018201a000249f018201a000249f018201a000249f01903e800011a000249f018201a000249f01903e800081a000242201a00067e2318760001011a000249f01903e800081a000249f01a0001b79818f7011a000249f0192710011a0002155e19052e011903e81a000249f01903e8011a000249f018201a000249f018201a000249f0182001011a000249f0011a000249f0041a000194af18f8011a000194af18f8011a0002377c190556011a0002bdea1901f1011a000249f018201a000249f018201a000249f018201a000249f018201a000249f018201a000249f018201a000242201a00067e23187600010119f04c192bd200011a000249f018201a000242201a00067e2318760001011a000242201a00067e2318760001011a0025cea81971f704001a000141bb041a000249f019138800011a000249f018201a000302590001011a000249f018201a000249f018201a000249f018201a000249f018201a000249f018201a000249f018201a000249f018201a00330da70101ff`

dataToHash = Buffer of (redeemerHex+datumHex+lanViewHex)

the resulting hash does not match the the one produced by the cli.

IMP to note - the script data hash matches the cli produced one when there are redeemers. Which can validate the implementation I have is correct when the redeemer is present. This is also confirmed by posting tx on chain resulting in a successful plutus script transaction.

**Comment by JaredCorduan** [comment]
`lanViewHex` should not have the Plutus V1 stuff in it. Since the transaction is using no Plutus scripts, the language views should be empty. Like with the empty redeemers, the correct thing to do would have been to have `lanViewHex` be the empty bytestring, but unfortunately it is the CBOR empty map, namey `A0`. So see if hashing `8081d87980A0` matches the CLI.

**Comment by JaredCorduan** [comment]
cc @rooooooooob (I'm not sure if this is the kind of wart you would also like to be aware of)
tl;dr - The serialized data used for the script integrity hash is a bit wonky in the case of the user supplying the optional datum for a newly created transaction output containing a datum hash, and only when no Plutus scripts are used. In particular, the bytes for the datums need to be prepended with `80` and appended with `A0`. This is sad, and we now need to decide if we want to change it in the next hard fork or not.

**Comment by ashisherc** [comment]
@JaredCorduan That is it. That worked. Thanks so much for looking into it, and for your time.

**Comment by JaredCorduan** [comment]
I am going to close this issues since I do no think it makes sense to change the behavior now.

We _could_ use the empty bytestring starting in conway (in the situation that there are no redeemers), but we would still have to support the current behavior in Alonzo and Babbage (and thus most libraries would probably have to support both behaviors). And there's another awkward point: how would we differentiate between 1) the transaction not including the redeemer field (in the CBOR) and 2) the transaction having set the redeemers to the empty map. So whatever we do, we are going to have to explain the tricky corners, so perhaps saying use `A0` is not so bad.
